### PR TITLE
Proposal for supporting deep copy of out parameters

### PR DESCRIPTION
**Pull request by mingweishih**
This PR adds the proposal of supporting deep copy of variable-length out parameters. The goal of the document is seeking the consensus on which design/direction that OE should take.

Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

### CODE REVIEW

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:15]:
```suggestion
An example of struct definitions is as follows.
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:71]:
```suggestion
The following code snippet serves as a common example that both designs aim to support.
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:79]: Incomplete code.  How is b->data freed?   Please include in the sample what the developer needs to do to free the blob.  e.g., is it
free(b->data);
?

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:96]: Line 29 talks about both directions (ECALL and OCALL) but here only talks about one direction (ECALL).
What about out parameters of OCALLs?

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:142]: Again this only talks about out parameters of ECALLs, what about out parameters of OCALLs?

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:149]:
```suggestion
  Note that the goal of `b_out` being `user_check` is to forward the original value
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:136]: Question: will the above also work for out parameters on _switchless_ calls?

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:80]: Updated.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:99]: Updated.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Good point. That's something we need to consider. Taking the switchless calls into account may further complicate the design without shared memory.

**Comment by mingweishih** [comment]
@HernanGatta I have a question regarding the shared memory support on OP-TEE. I do see something in the document: https://optee.readthedocs.io/en/latest/architecture/core.html#shared-memory. Can you add more on this? If shared memory is available on OP-TEE as well, we could just go with the simpler design/implementation.

**Comment by HernanGatta** [comment]
> @HernanGatta I have a question regarding the shared memory support on OP-TEE. I do see something in the document: https://optee.readthedocs.io/en/latest/architecture/core.html#shared-memory. Can you add more on this? If shared memory is available on OP-TEE as well, we could just go with the simpler design/implementation.

Hello Ming-Wei, thank you for tagging me in this. I'm thinking it through and trying to see what the best way to go about this is on OP-TEE.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (radhikaj) [reply docs/DesignDocs/DeepCopyOutParameters.md]: switchless calls are not supported on OPTEE.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:178]: why is there no free(b->buf); ?

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:93]: We can free the allocated buffer automatically (in the oeedger8r-generated code). However, this may not work with 
 a corner case: instead of allocating memory, the callee assigns a global pointer to the `b->buf`. See the following example.
```
/* Managed somewhere else */
char *g_buf;
size_t g_buf_len;

void foo(Blob *b)
{
    b->len = g_buf_len;
    b->buf = g_buf;
}
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:137]: Thanks for the suggestion. But it's unclear how memdup supports the copy of nested pointers, which I need to think.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:178]: See my previous comment regarding the special case.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [reply docs/DesignDocs/DeepCopyOutParameters.md:93]: I don't understand what you're proposing as an answer.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (bodzhang) [inline docs/DesignDocs/DeepCopyOutParameters.md:211]: If higher level code did not verify that `_pargs_out->b->buf`, with length of `b->len`, is fully outside of the enclave memory already, the code here must verify that before performing the copy.

Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:211]: `b->len` is inside the enclave while `_pargs_out->b->len` is on the hosot.

**Review by anakrish — COMMENTED** [review]
Thanks for investigating this complex problem, writing up the design and seeking consensus.

It looks good to me. However, I'd like to see the following addressed:

1) One of the key concerns raised during security review of deep-copy was that objects ought to be transferred in a single memcpy across the host/enclave boundary. Copying an object tree, object by object is known to be error prone and hard to get right.
The current proposal seems to perform a malloc for every object, then copy it over, then malloc and copy the child objects. Can this be changed so that objects are copied across the host/enclave security boundary in a single memcpy and then reconstructed at the other side.
This is the approach edger8r currently takes prior to this proposal.


2) Variable length outputs can be encoded as two calls where the first call returns the size of the variable length output and the second call returns the object. 
    For example, the unsupported edge function
    ```c
    void get_data([out]size_t* size, [out] uint8_t** data);
    ```
    can be expressed as the following that works in both SGX and OPTEE
    ```c
     void get_data_size([out]size_t* size);
     void get_data(size_t size, [out,size=size] uint8_t* data);
    ```
One can envision a code generation scheme where the edger8r auto-generates and uses two calls in the same direction to implement deep-copy out parameters. 
The proposal does not clarify whether such an approach was considered or not.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:629]: The design has been modified and hence the comments are no longer applicable.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Added `oe_allocate_host_buffer` and `oe_free_host_buffer`.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated design.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated the example.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated the doc.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated.

**Comment by mingweishih** [comment]
Thanks for the review, I've updated the both the design and the documentation.

> Thanks for investigating this complex problem, writing up the design and seeking consensus.
> 
> It looks good to me. However, I'd like to see the following addressed:
> 
> 1. One of the key concerns raised during security review of deep-copy was that objects ought to be transferred in a single memcpy across the host/enclave boundary. Copying an object tree, object by object is known to be error prone and hard to get right.
>    The current proposal seems to perform a malloc for every object, then copy it over, then malloc and copy the child objects. Can this be changed so that objects are copied across the host/enclave security boundary in a single memcpy and then reconstructed at the other side.
>    This is the approach edger8r currently takes prior to this proposal.

The updated design now uses single memcpy across the boundary. More specifically,
- The caller serializes the out parameter into a local buffer.
- The enclave performs a single memcpy that copies the buffer across the boundary.
- The callee re-constructs the out parameter.

> 2. Variable length outputs can be encoded as two calls where the first call returns the size of the variable length output and the second call returns the object.
>    For example, the unsupported edge function
>    ```c
>    void get_data([out]size_t* size, [out] uint8_t** data);
>    ```
>    
>    
>    can be expressed as the following that works in both SGX and OPTEE
>    ```c
>     void get_data_size([out]size_t* size);
>     void get_data(size_t size, [out,size=size] uint8_t* data);
>    ```
> 
> One can envision a code generation scheme where the edger8r auto-generates and uses two calls in the same direction to implement deep-copy out parameters.
> The proposal does not clarify whether such an approach was considered or not.

The new design uses the similar strategy but avoids using an additional ECALL/OCALL.
- The buffer size is calculated by the caller based on the `_output_bytes_written` that is set and returned by the callee.
- Introduce a `_host_buffer` to the internal `args_t` struct that is used to pass the host pointer.

I add the possible design for the double OCALL/ECALL scheme you suggested in the discussion section.
However, I would still prefer to implement the current design first and revisit this option later.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (anakrish) [inline docs/DesignDocs/DeepCopyOutParameters.md:157]: You could also add a `_host_buffer_size` member. The first 3 fields of any marshalling struct would then be:
```c
typedef struct _fcn_args_t
  {
      oe_result_t _result;
      uint8_t* _host_buffer;
      size_t _host_buffer_size;
      ... // other marshaling fields
  }
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (anakrish) [inline docs/DesignDocs/DeepCopyOutParameters.md:316]: This does not have to be done by the generated code. 
The generated code just needs to do 
```c
    args->_host_buffer = _deepcopy_out_buffer;
    args->_host_buffer_size = _deepcopy_out_buffer_size;
```

The ECALL dispatching infrastructure can determine the best way to marshal _host_buffer over to the host, according to the platform (SGX/OPTEE).

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (anakrish) [inline docs/DesignDocs/DeepCopyOutParameters.md:407]: Instead of this, we could generate:
```
pargs_out->_host_buffer = _deepcopy_out_buffer;
pargs_out->_host_buffer_size = deepcopy_out_buffer_size;
```

The infrastructure can marshal this over to the host as appropriate. The generated code doesn't have to deal with host memory.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (anakrish) [inline docs/DesignDocs/DeepCopyOutParameters.md:463]: As for ecall, add `_host_buffer_size`.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (anakrish) [inline docs/DesignDocs/DeepCopyOutParameters.md:699]: Once the earlier suggestions for ecall are incorporated, the generated code for ecall/ocall deep-copy out parameters will look identical.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (anakrish) [inline docs/DesignDocs/DeepCopyOutParameters.md:742]: By moving host buffer allocation to the infrastructure rather than the generated code, there is no longer any reliance on edger8r generated code on shared memory. It is the infrastructure that needs to marshal the deep-copy parameter appropriately on each platform (SGX, OPTEE).

**Review by anakrish — APPROVED** [review]
Looks great!

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated. Currently, the enclave (OCALL wrapper) performs an additional check against the buffer to ensure it resides in enclave memory.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Removed the section.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated.

**Comment by radhikaj** [comment]
@dthaler, this is an experimental feature and follows the proposal we discussed in SIF-Arch last November. Would it be possible for you to take another look?

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: The design has been updated such that the memory transmission is now handled internally, not by the oeedger8r-generated code. Please take a look.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Now the callee-allocated memory will be freed automatically. Please check the updated document.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Updated the proposal to mention that the design does not change the behavior of existing `inout` parameter.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:12]: Recommend adding a reference to full_edl_serialization.md, and updating full_edl_serialization.md with a reference to this design doc.  Also full_edl_serialization.md includes the statement:

> Although EDL provides sufficient mechanisms for performing full serialization, the generator does not currently enforce these mechanisms. By default, the generator produces edge routines that implicitly copy untrusted memory references into the enclave. This results in potential security flaws and non-portable code.
> 
> To overcome these problems, the generator can be modified to warn when an EDL specification cannot be fully serialized.

I'd suggest updating that text in this PR.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:35]:
```suggestion
To pass the `NestedBlob` struct in an ECALL/OCALL, oeedger8r currently supports the following two cases:
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:83]:
```suggestion
  The caller, either an enclave or a host, is required to prepare an instance of `NestedBlob`.
  In this example, `nb` is the instance of `NestedBlob`. The caller then passes `nb` to
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:146]:
```suggestion
oeedger8r-generated code serializes the content of `nb` into a local buffer. Next, the content of the buffer
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:153]:
```suggestion
Note that this design is compatible with the switchless call feature, which also requires shared memory.
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:157]:
```suggestion
  The design adds `deepcopy_out_buffer` and `deepcopy_out_buffer_size` members to every internal struct of ECALLs.
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:160]:
```suggestion
  between the host and the enclave. The values (zero or not) of the two members are also used by the
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:180]:
```suggestion
    
The following is an example of oeedger8r-generated code:
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:439]:
```suggestion
  Similar to the case of ECALL, the design adds the same modifications to every internal struct of OCALLs.
```

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:713]:
```suggestion
in SGX while not introducing breaking changes to the OE's support on OP-TEE.
```

**Review by dthaler — APPROVED** [review]
Design looks good to me, just some comments.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (dthaler) [inline docs/DesignDocs/DeepCopyOutParameters.md:713]:
```suggestion
in SGX while not introducing breaking changes to the OE's support on the OP-TEE.
```
@HernanGatta is there anything in here that would break something that works on OP-TEE?  If not, then this doesn't need to be experimental.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md:12]: Updated. Please take a look.

File: docs/DesignDocs/DeepCopyOutParameters.md
Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: The design won't break existing OP-TEE ECALLs/OCALLs that don't use the deep-copy out annotations (there shouldn't be any).
Even if developers try to use deep-copy out on OP-TEE, the compilation would go thru but the code just won't work unless we implement the OP-TEE support internally.

Given that we currently don't specify target TEE to the oeedger8r tool, we don't have a good way to tell developers that deep-copy out is not supported on OP-TEE. As a result, the current plan is guarding the feature with the `--experimental` flag, which makes the feature implicit. If the flag is not set, the developer will see an error message like:
```
`out' for user defined type `MyStruct\\*' requires the --experimental option`.
```

@dthaler @anakrish Do you think we can make this as non-experimental? We do have the switchless call feature that is similar to this case (only works for SGX).

Alternatively, we could introduce a TEE flag to oeedger8r or let the tool to detect the target based on the platform (x86 or arm). We could also add the OP-TEE support, which requires @HernanGatta's inputs. I think both of them could be future work.

Comment (anakrish) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Let's first merge this in as experimental, and get feedback from the user that wants this.

Comment (dthaler) [reply docs/DesignDocs/DeepCopyOutParameters.md]: Unless we know of things that would break OP-TEE, I don't think OP-TEE should be any reason to make this be experimental.  If we think it needs to be experimental for SGX because SGX might break, then ok.  But I'd probably argue for non-experimental.

Comment (mingweishih) [reply docs/DesignDocs/DeepCopyOutParameters.md]: We're thinking of following the previous path on deep-copy support: making the feature as experimental first and then get feedback from users/ask for security reviews before making this as non-experimental.

Comment (radhikaj) [reply docs/DesignDocs/DeepCopyOutParameters.md]: I agree with what Ming-Wei said. Let’s get feedback and see if some folks can look at the security implications. Cc @mrragava

**Comment by radhikaj** [comment]
I am merging this since it has 2 approvals. A move to non experimental can happen in a different PR

**Pull request merged**
